utils/memory_bench/src/main.rs

```diff
@@ -0,0 +1,421 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+use quote::quote;
+use serde_json::json;
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::PathBuf;
+use std::process;
+use std::process::Command;
+use std::{env, process::Stdio};
+use std::{fs, io::BufReader};
+
+fn inject_dhat_scoped_variable(syntax: &mut syn::File) {
+    let dhat = syn::parse_str("let _dhat = Dhat::start_heap_profiling();")
+        .expect("Unable to parse the dhat string");
+
+    // Find the main function.
+    for item in syntax.items.iter_mut() {
+        if let syn::Item::Fn(ref mut fn_item) = item {
+            if fn_item.sig.ident == "main" {
+                // Create a new vector with the injected statement at the beginning.
+                let mut new_stmts = vec![dhat];
+                for stmt in fn_item.block.stmts.drain(..) {
+                    new_stmts.push(stmt);
+                }
+                fn_item.block.stmts = new_stmts;
+                return;
+            }
+        }
+    }
+
+    panic!("Unable to find the main function.");
+}
+
+fn inject_allocator_declaration(syntax: &mut syn::File) {
+    let use_code = "use dhat::{Dhat, DhatAlloc};";
+    let allocator_code = "
+        #[global_allocator]
+        static ALLOCATOR: DhatAlloc = DhatAlloc;
+    ";
+    let mut new_items = vec![
+        syn::parse_str(use_code).expect("Unable to parse the dhat use string"),
+        syn::parse_str(allocator_code).expect("Unable to parse the allocator string"),
+    ];
+
+    for item in syntax.items.drain(..) {
+        new_items.push(item);
+    }
+    syntax.items = new_items;
+}
+
+/// dhat requires that the files are manually instrumented. This function automates
+/// that process, by processing the AST using the syn package, modifying the AST,
+/// and then finally writing it back out to a file with a slightly different name.
+/// The instrumented version of the file can then be run.
+fn run_dhat_injection(filename: &PathBuf) -> String {
+    let mut file = File::open(&filename).expect("Unable to open file");
+
+    let mut src = String::new();
+    file.read_to_string(&mut src).expect("Unable to read file");
+
+    let mut syntax = syn::parse_file(&src).expect("Unable to parse file");
+
+    inject_allocator_declaration(&mut syntax);
+    inject_dhat_scoped_variable(&mut syntax);
+
+    quote!(#syntax).to_string()
+}
+
+fn print_usage() {
+    eprintln!("Usage: cargo run --component memory -- [os] [component] [...example]");
+    eprintln!("e.g. : cargo run --component memory -- macos-latest datetime work_log skeletons");
```

Comment:
Given that you can compile examples irrespective of the component (e.g., `cargo run --example work_log` from the top level package), it would be nice if we can avoid putting the component in the command line argument.  It also makes it harder to scale to new components and examples.

Reply:
I did this since I was concerned there would be multiple examples named the same thing. I'm not sure what happens then. If we only use example names, then I'm curious if we should change the directory structure as well for the charts.

It would be much simpler in the implementation to iterate over a single list of example names.

Reply:
I looked into this again and you can run `cargo run -p icu_datetime --example work_log`. However, if we still want to rewrite the example file, cargo doesn't provide a mechanism to get the path to the example .rs file (as far as I know).

Reply:
You could get the path programmatically by parsing `.toml` files and listing `/examples` dirs, but that's additional complexity.

Reply:
> I did this since I was concerned there would be multiple examples named the same thing.

I think we can assume this invariant. We should work to prevent example names from conflicting.

> I'm curious if we should change the directory structure as well for the charts.

I'm in favor of treating examples as being mostly flat.

Reply:
It looks like `cargo run` will return an error code if there is an example run from the root with the duplicates, so the CI job should fail. This invariant should be enforced then 👍

> error: `cargo run` can run at most one executable, but multiple were specified

Reply:
> Given that you can compile examples irrespective of the component (e.g., cargo run --example work_log from the top level package), it would be nice if we can avoid putting the component in the command line argument. It also makes it harder to scale to new components and examples.

So it turns out that features require the path to the Cargo.toml and can't be specified in a virtual workspace. To get around this I ended up specifying a `{package}/{example}` syntax for the tool. I then used the cargo metadata information to locate the Cargo.toml file from the package name.